### PR TITLE
Add start, end, progress, and error callbacks for ESP8266HTTPUpdate

### PR DESCRIPTION
**Pull request by earlephilhower**
Replaces abandoned #1817

Add an optional std::function callback (so it supports lambdas and
normal functions) with an optional data parameter (for passing in a
`this` or other state info).

Update example with its use.

From @baruch's original pull request:
The callback is called when the upgrade actually starts rather than just
the initial query so that the user can know that it will not take longer
and can also prepare for the upgrade by shutting down other works.

### CODE REVIEW

File: libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.h [hunk 1]
```diff
@@ -64,6 +64,8 @@ enum HTTPUpdateResult {
 
 typedef HTTPUpdateResult t_httpUpdate_return; // backward compatibility
 
+typedef std::function<void(void *data)> HTTPUpdateCB;
```
Comment (devyte) [inline libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.h]: prefer using = syntax

**Comment by earlephilhower** [comment]
I'm going to merge this with an update for #2694, too, which has add'l CBs.

File: libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.h (hunk 1 above)
Comment (earlephilhower) [reply libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.h]: Will do

**Comment by earlephilhower** [comment]
Now implements the 4 CBs in@karlp's #2694.

**Comment by earlephilhower** [comment]
Fixes #4183

**Pull request merged**